javalin/src/main/java/io/javalin/core/validation/Validator.kt

```diff
@@ -6,107 +6,44 @@
 
 package io.javalin.core.validation
 
+import io.javalin.core.util.JavalinLogger
 import io.javalin.http.BadRequestResponse
+import io.javalin.http.InternalServerErrorResponse
 
-open class Validator<T>(val value: T?, val messagePrefix: String = "Value", val key: String = "Parameter") {
+enum class ValidationError { NULLCHECK_FAILED, TYPE_CONVERSION_FAILED, CUSTOM_CHECK_FAILED, DESERIALIZATION_FAILED }
 
-    data class Rule<T>(val fieldName: String, val test: (T) -> Boolean, val invalidMessage: String)
-
-    protected open var errors: MutableMap<String, MutableList<String>>? = null
-
-    protected val rules = mutableSetOf<Rule<T>>()
-
-    @JvmOverloads
-    open fun check(predicate: (T) -> Boolean, errorMessage: String = "Failed check"): Validator<T> {
-        rules.add(Rule(key, predicate, errorMessage))
-        return this
-    }
-
-    //These two options will fail fast but only provide the first failure.
-    fun get(): T = getOrNull() ?: throw BadRequestResponse("$messagePrefix cannot be null or empty")
+/**
+ * The non-nullable [Validator] uses [Rule] rules, but checks if value is null before calling them.
+ * The [check] method wraps its non-nullable predicate in a nullable predicate
+ */
+open class Validator<T>(value: T?, val fieldName: String) : BaseValidator<T>(value) {
 
-    fun getOrNull(): T? {
-        if (value == null) return null
-        return rules.find { !it.test.invoke(value) }?.let { throw BadRequestResponse("$messagePrefix invalid - ${it.invalidMessage}") } ?: value
+    init {
+        addRule(fieldName, { it != null }, {
+            JavalinLogger.info("Parameter '${fieldName}' cannot be null")
+            ValidationError.NULLCHECK_FAILED.name
+        })
     }
 
-    open fun isValid(): Boolean {
-        if (errors == null) {
-            validate()
-        }
-        return errors!!.isEmpty()
-    }
+    fun allowNullable() = NullableValidator(value, fieldName)
```

Comment:
this isn't inheriting any rules it already has, which makes things a little confusing with the builder pattern potentially.
```
validator.check(...).allowNullable().check(...)
```

It seems a bit unexpected that only the second check would run (even if it's a bit strange for someone to do it in that order). Of course, if you end up overriding get like I suggested previously, then you'll end up with null pointers due to the way that the check is passing the value here.

Reply:
This is one of the tradeoffs of this slightly hacky approach. We could throw an illegal state exception if rules have been added?